packages/cli/src/plugin.js

```diff
@@ -0,0 +1,228 @@
+    // Do nothing if there are no enabled plugins for this namespace
+    if (namespacedPlugins.size === 0) {
+      continue
+    }
+
+    // For help output we only show the root level commands which doesn't actually need
+    // to load any plugins other than the @redwoodjs namespaced ones
+    if (showingRootHelp && namespace !== '@redwoodjs') {
+      yargs.command({
+        command: `${namespace} <command>`,
+        describe: `${namespace} plugin commands`,
+        builder: () => {},
+        handler: () => {},
+      })
+      continue
+    }
+
+    const namespacedCommands = []
+    // Load all plugins for this namespace
+    for (const namespacedPlugin of namespacedPlugins) {
+      // Check the cache to see if we can skip loading this plugin
+      // If this plugin doesn't have a command that matches the first word we can skip loading it
+      const cacheEntry = pluginCommandCache[namespacedPlugin.package]
+      if (
+        !showingNamespaceHelp &&
+        cacheEntry &&
+        !cacheEntry.includes(firstWord)
+      ) {
+        continue
+      }
+
+      let plugin
+      try {
+        plugin = await import(namespacedPlugin.package)
+      } catch (error) {
+        // TODO: Batch all missing plugins and install them in one go
+        if (error.code === 'MODULE_NOT_FOUND') {
+          if (!autoInstall) {
+            console.warn(
+              chalk.yellow(
+                `⚠️  Plugin "${namespacedPlugin.package}" cannot be loaded because it is not installed and "autoInstall" is disabled.`
+              )
+            )
+            continue
+          }
+          console.log(
+            chalk.green(`Installing plugin "${namespacedPlugin.package}"...`)
+          )
+          // Install the plugin
+          await installModule(
+            namespacedPlugin.package,
+            namespacedPlugin.version,
+            true
+          )
+          plugin = await import(namespacedPlugin.package)
+        }
+      }
+
+      if (!plugin) {
+        console.log(
+          chalk.red(`❌  Plugin "${namespacedPlugin.package}" failed to load.`)
+        )
+        continue
+      }
+
+      // Add the plugin to the cache entry
+      pluginCommandCache[namespacedPlugin.package] = []
+      for (const command of plugin.commands) {
+        // Add the first word of the command to the cache entry
+        pluginCommandCache[namespacedPlugin.package].push(
+          command.command.split(' ')[0]
+        )
+        // Add any aliases of the command to the cache entry
+        pluginCommandCache[namespacedPlugin.package].push(
+          ...(command.aliases || [])
+        )
+      }
+
+      // Add these commands to the namespace list
+      namespacedCommands.push(...plugin.commands)
+    }
+
+    // If we didn't load any commands for this namespace we can skip registering it
+    // unless we're loading all namespaces for the help output
+    if (namespacedCommands.length === 0 && !showingRootHelp) {
+      continue
+    }
```

Comment:
@Josh-Walker-GM could you help me understand when this condition would be met?

Reply:
Right now this occurs when we've filtered out all possible commands. This might happen for example if we could tell from the cache that no plugins could possibly satisfy the command you entered. 

This raises the point that the help message is not great in that scenario I give above. We didn't load the plugins into memory because we know none can satisfy the user but as a result we don't get a very good help message. For example if I run `yarn rw @joshgmw testing` - which no `@joshgmw` plugin provides then I get the following as the last line of the output:
```
Unknown arguments: @joshgmw, testing
```
Maybe we should in that case load them all into memory and yargs will then produce the appropriate help message. This would take a little refactor to loop round again and load all the plugins. In the meantime I have added a console warning pointing the user to run `yarn rw @namespace --help`

Reply:
I'll refactor this later tonight when I get some more time so that it just loads all the plugins and provides the standard yargs help.